evm/src/cpu/contextops.rs

```diff
@@ -1,18 +1,77 @@
+use itertools::izip;
 use plonky2::field::extension::Extendable;
 use plonky2::field::packed::PackedField;
 use plonky2::field::types::Field;
 use plonky2::hash::hash_types::RichField;
 use plonky2::iop::ext_target::ExtensionTarget;
 use plonky2::plonk::circuit_builder::CircuitBuilder;
 
+use super::columns::ops::OpsColumnsView;
 use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
 use crate::cpu::columns::CpuColumnsView;
 use crate::cpu::kernel::constants::context_metadata::ContextMetadata;
 use crate::cpu::membus::NUM_GP_CHANNELS;
 use crate::memory::segments::Segment;
 
+// If true, the instruction will keep the current context for the next row.
+// If false, next row's context is handled manually.
+const KEEPS_CONTEXT: OpsColumnsView<bool> = OpsColumnsView {
+    binary_op: true,
+    ternary_op: true,
+    fp254_op: true,
+    eq_iszero: true,
+    logic_op: true,
+    not: true,
+    shift: true,
+    keccak_general: true,
+    prover_input: true,
+    pop: true,
+    jumps: true,
+    pc: true,
+    jumpdest: true,
+    push0: true,
+    push: true,
+    dup: true,
+    swap: true,
+    context_op: false,
+    mstore_32bytes: true,
+    mload_32bytes: true,
+    exit_kernel: true,
+    m_op_general: true,
+    syscall: true,
+    exception: true,
+};
+
+fn eval_packed_keep<P: PackedField>(
+    lv: &CpuColumnsView<P>,
+    nv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    for (op, keeps_context) in izip!(lv.op.into_iter(), KEEPS_CONTEXT.into_iter()) {
```

Comment:
`KEEPS_CONTEXT` is an `OpsColumnsView<T>` which would not be a single item when turned into an iterator (it would have `NUM_OPS_COLUMNS` elements). This is why Hamy's using it to map value of an op flag (which can't be checked directly with an `if/else` condition because of constraints, to the expected behavior of this flag.

Although not needed, `izip!` has been used in other places of the codebase so it doesn't really bothers me.

Reply:
I'm not a fan either of the `OpsColumnView` iterator approach they took (which is replicated in several other places in the EVM), but I don't see a better way to do it.

For `izip!` I was indeed mimicking what they do in other places, like in `stack.rs`.